### PR TITLE
feat: add optional contribution option

### PR DESCRIPTION
**Pull request by wpf500**
Added the optional contribution check box to the join page and the setting to enable it to the membership builder.

I also fixed the long standing #90 which was only reported in Jan 2021 :wink:, and abstracted the contribution period/amount selector into a new generic `AppChoice`.

### Screenshots

image

## Checklist before requesting a review

- [x] Done a self-review of my code
- [x] Run `npm run check` and addressed any problems
- [x] PR doesn't have merge conflicts

## Checklist before merging

- [x] Translations for all new i18n strings

### CODE REVIEW

File: src/components/forms/AppInput.vue [hunk 1]
```diff
@@ -80,6 +83,7 @@ const props = withDefaults(
     sameAs: undefined,
     pattern: undefined,
     prefix: undefined,
+    suffix: undefined,
```
Comment (JumpLink) [inline src/components/forms/AppInput.vue:86]: Why are the values here initialised with `undefined`?

File: src/components/contribution/ContributionAmount.vue [hunk 2]
```diff
@@ -116,7 +118,9 @@ const hasError = computed(() => validation.value.$errors.length > 0);
 
 const rules = computed(() => ({
   amount: {
-    minValue: minValue(toRefs(props).minAmount),
+    ...(!props.disabled && {
+      minValue: minValue(toRefs(props).minAmount),
+    }),
   },
 }));
```
Comment (JumpLink) [inline src/components/contribution/ContributionAmount.vue]: I find this difficult to read, why not instead like this?

```ts
cconst ruls = computed(() => {
  if (props.disabled) return { amount: {} };

  return {
    amount: {
      minValue: minValue(toRefs(props).minAmount),
    },
  };
});
```

**Review by JumpLink — APPROVED** [review]
is approved, just two beauty questions / suggestions

File: src/components/forms/AppInput.vue (hunk 1 above)
Comment (wpf500) [reply src/components/forms/AppInput.vue:86]: It seems strange to me too but if I don't set a default on optional props then I get an ESLint warning:

image

File: src/components/contribution/ContributionAmount.vue (hunk 2 above)
Comment (wpf500) [reply src/components/contribution/ContributionAmount.vue]: I'm definitely guilty of doing this sort of thing way too much :+1:

What do you think of using a ternary operator, e.g.
```ts
const rules = computed(() =>
  props.disabled
    ? { amount: {} }
    : {
        amount: {
          minValue: minValue(toRefs(props).minAmount),
        },
      }
);
```

I find ternary operators easy to read and offer less potential for future developers to do strange things with not seeing early returns etc.

Comment (JumpLink) [reply src/components/contribution/ContributionAmount.vue]: Looks good 👍

File: src/components/forms/AppInput.vue (hunk 1 above)
Comment (JumpLink) [reply src/components/forms/AppInput.vue:86]: okay

**Pull request merged**